pkg/policy/selectorcache.go

```diff
@@ -0,0 +1,336 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"sort"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/sirupsen/logrus"
+)
+
+// CachedSelector represents an identity selector owned by the selector cache
+type CachedSelector interface {
+	// GetSelections returns the cached set of numeric identities
+	// selected by the CachedSelector.  The retuned slice must NOT
+	// be modified, as it is shared among multiple users.
+	GetSelections() []identity.NumericIdentity
+
+	// String returns the string representation of this selector.
+	// Used as a map key.
+	String() string
+}
+
+// CachedSelectionUser inserts selectors into the cache and gets update
+// callbacks whenever the set of selected numeric identities change for
+// the CachedSelectors pushed by it.
+type CachedSelectionUser interface {
+	// IdentitySelectionUpdated implementations MUST NOT call back
+	// to selector cache while executing this function!
```

Comment:
To avoid recirsive/double locking. This function (callback) is called with the cache lock held, and calling any selector cache "API" function will panic.